core/src/main/java/org/mule/runtime/core/internal/policy/DefaultPolicyManager.java

```diff
@@ -207,6 +212,41 @@ public void initialise() throws InitialisationException {
     policyPointcutParametersManager =
         new PolicyPointcutParametersManager(registry.lookupAllByType(SourcePolicyPointcutParametersFactory.class),
                                             registry.lookupAllByType(OperationPolicyPointcutParametersFactory.class));
+
+    // Register flow disposal listener
+    muleContext.getNotificationManager().addListener(new FlowConstructNotificationListener<FlowConstructNotification>() {
+
+      @Override
+      public boolean isBlocking() {
+        return false;
+      }
+
+      @Override
+      public void onNotification(FlowConstructNotification notification) {
+        if (Integer.parseInt(notification.getAction().getIdentifier()) == FLOW_CONSTRUCT_DISPOSED) {
+          LOGGER.debug("Invalidating flow from caches named {}", notification.getResourceIdentifier());
+          // Invalidate flow from caches
+          invalidateDisposedFlowFromCaches(notification.getResourceIdentifier());
+        }
+      }
+    });
+  }
+
+  private void invalidateDisposedFlowFromCaches(String flowName) {
+    // Invalidate from outer "with policy cache"
+    sourcePolicyOuterCache.asMap().keySet().stream()
+        .filter(pair -> pair.getFirst().equals(flowName))
+        .forEach(matchingPair -> sourcePolicyOuterCache.invalidate(matchingPair));
+
+    // Invalidate from inner "with policy cache"
+    sourcePolicyInnerCache.asMap().keySet().stream()
+        .filter(pair -> pair.getFirst().equals(flowName))
+        .forEach(matchingPair -> sourcePolicyInnerCache.invalidate(matchingPair));
+
+    // Invalidate from "no policy cache"
+    noPolicySourceInstances.asMap().keySet().stream()
```

Comment:
you can use just `invalidate(key)` for this instead of `asMap` and the rest of the stuff